### PR TITLE
ext_authz: allows multiple headers of same name in Denied response

### LINKED ISSUE #8649: ext_authz filter should allow multiple `Set-Cookie` headers to be added by the `Check` response
*Title*: ext_authz filter should allow multiple `Set-Cookie` headers to be added by the `Check` response

*Description*:

We are writing an implementation of the [ext_authz](https://github.com/envoyproxy/envoy/tree/master/test/extensions/filters/http/ext_authz)'s `Check` interface. Our implementation would like to return a `Denied` status, set the http response to be a 302, and add multiple `Set-Cookie` headers to the http response to manipulate multiple cookies.

As additional background, it is not possible in many browsers to manipulate multiple cookies in a single `Set-Cookie` header because it is disallowed by [Section 3 of RFC 6265](https://tools.ietf.org/html/rfc6265#section-3):

> Origin servers SHOULD NOT fold multiple Set-Cookie header fields into a single header field.

As a result, when given a `Set-Cookie` header of the form `Set-Cookie: a=1, b=2`, modern browsers will ignore the `b` cookie.  Instead, they need the header to be repeated like this:
```
Set-Cookie: a=1
Set-Cookie: b=2
```

So, we would like to add multiple `Set-Cookie` headers, but it seems that the `ext_authz` filter is preventing this.

Today, when multiple headers of the same name are added by the `Check` response, the ext_authz filter always only adds the last one.  You can see the relevant code in [ext_authz.cc](https://github.com/envoyproxy/envoy/blob/6def4a4a4cecdd1f54024cea6084a2bc1d9ef83d/source/extensions/filters/http/ext_authz/ext_authz.cc#L211).

It seems like the `response_headers.remove(header.first)` line is intended to satisfy the requirement described by the [OverrideEncodingHeaders unit test](https://github.com/envoyproxy/envoy/blob/6def4a4a4cecdd1f54024cea6084a2bc1d9ef83d/test/extensions/filters/http/ext_authz/ext_authz_test.cc#L1171).  However, that line of code also seems to cause it to ignore the `Check` response's intention to add multiple headers of the same name.

*Proposed Solution*:

One possible solution would be for the `ext_authz` filter to ensure that headers from the `Check` response do override existing headers, while also allowing the `Check` response to override a single existing header with multiple headers of the same name.  For example, this could be implemented in two passes: first, call `remove` on all of the headers set by the `Check` response, and then in a second pass add all of the headers set by the `Check` response. The [`addCopy`](https://github.com/envoyproxy/envoy/blob/6def4a4a4cecdd1f54024cea6084a2bc1d9ef83d/source/common/http/header_map_impl.cc#L427) method already has the smarts built-in to decide if a header with multiple values should be folded (combined into a single header with multiple values separated by commas) or should be repeated as separate headers of the same name. Specifically, the `addCopy` method is already smart enough to avoid folding `Set-Cookie` headers, which is exactly what we need.

*Relevant Links*:

- [There was a similar issue filed against Envoy](https://github.com/envoyproxy/envoy/issues/7488), but it was determined that Envoy was correctly allowing multiple `Set-Cookie` headers to be added.
- [There was also a similar issue in Istio Mixer](https://github.com/istio/istio/pull/15581) and that fix was merged.

### PR DESCRIPTION
**Pull request by cfryanr**
Description:

Enhance the ext_authz filter to allow multiple `Set-Cookie`
headers to be added by a `Denied` `Check` response.

Previously, when the `Check` response contained multiple
headers of the same name, only the last one would be applied
in the http response. Please see full description of problem
in #8649.

Risk Level: Low
Testing: Unit test
Docs Changes: N/A
Release Notes: N/A

Fixes #8649

### CODE REVIEW

**Comment by srwaggon** [comment]
@dio, we believe the test failures might be unrelated to our changeset.

Could you re-run the test or give us some feedback about why these tests might be failing?

**Comment by peterhaochen47** [comment]
@lizan We believe the test failures are unrelated. Could you take a look as well?

**Comment by dio** [comment]
@srwaggon @peterhaochen47 can you try to merge master? Thanks!

**Comment by srwaggon** [comment]
@dio Thanks for the advice 😄 That did it! This PR looks like it can be merged now with an approval!

File: source/extensions/filters/http/ext_authz/ext_authz.cc [hunk 1]
```diff
@@ -212,8 +212,10 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
           ENVOY_STREAM_LOG(trace,
                            "ext_authz filter added header(s) to the local response:", callbacks);
           for (const auto& header : headers) {
-            ENVOY_STREAM_LOG(trace, " '{}':'{}'", callbacks, header.first.get(), header.second);
             response_headers.remove(header.first);
+          }
+          for (const auto& header : headers) {
```
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc]: Yeah, the requirement of these two-passes is not obvious at first glance. Could you add comments around this? Transferring the proposed solution text in https://github.com/envoyproxy/envoy/issues/8649 should be good. But whatever explains why we're doing this helps, up to you.

Comment (cfryanr) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Thanks for the feedback, @dio. I pushed another commit to add the suggested comments. Please take a look.

**Comment by dio** [comment]
@gsagula could you help to verify this as well? Thanks!

**Comment by cfryanr** [comment]
Hi @dio and @gsagula, 

Thanks for reviewing the PR. Is it ready to be accepted? Is there anything else that we can do to help it get accepted?

I ask because it is blocking the initial release of the new istio-ecosystem [authservice](https://github.com/istio-ecosystem/authservice) project, which needs to be able to set and delete multiple cookies.

Thanks!

**Comment by dio** [comment]
@cfryanr seems like you need to merge master again. Sorry.

**Comment by cfryanr** [comment]
Hi @dio. Thanks. I merged master into the fork. There were no merge conflicts.

**Pull request merged**